### ISSUE TITLE
GPX import / export

### ISSUE THREAD

**Issue by zltnzltn** [traccar/traccar#958]
Hi, 

this one is a feature request.

I have 2-3 years of tracks, on everytrial.com and id like to import those gpx files into my traccar server (mysql).

An import/export to/into gpx file feature would be nice.

or a how to about importing manually gpx files into the mysql database.

Thank you

**Comment by renaudallard** [comment]
Export would be a really nice feature, it is/was available in vitalidze version of traccar, is there any way you could merge that code?

**Comment by tananaev** [comment]
It's not possible to merge the code because it's completely different from what I plan to implement.

**Comment by renaudallard** [comment]
So there is a plan for that feature, nice. We are already further :)

**Comment by tananaev** [comment]
Yes, it's definitely in plans. It's one of the highest priority features.

**Comment by Anton-V-K** [comment]
Any chance the GPX export will be available soon? :)

**Comment by tananaev** [comment]
It actually has been available for a while, but only through API. There is no UI for it at the moment.

**Comment by Anton-V-K** [comment]
Thanks for the hint. I've found the needed API request ([GET /reports/route](https://www.traccar.org/api-reference/#path--reports-route)), but I'm not able to make it work through simple request like `http://demo.traccar.org/api/reports/route?deviceId=253999&from=2017-12-01T00:00:00Z&to=2017-12-31T00:00:00Z`... Is it supposed to work this way?

**Comment by tananaev** [comment]
No, you need to include correct Accept header in the request.

**Comment by Anton-V-K** [comment]
Thanks for the reply.
It all makes usage quite complicated... Until the user-friendly export is ready, can we have the export with REST API in a "simpler mode", please? :)
Anyway I've tried to use `curl` this way:
```
curl -u myemail -H "Accept: application/json" "http://demo.traccar.org/api/reports/route?deviceId=253999&from=2017-12-01T00:00:00Z&to=2017-12-31T00:00:00Z"
```
The server responds with a kind of error: `Manager access required - SecurityException (PermissionsManager:159 < *:289 < Route:45 < ReportResource:44 < ...)`
Have I missed something?

**Comment by tananaev** [comment]
The device id is almost certainly wrong one. It should be internal id.

**Comment by Anton-V-K** [comment]
Ah, good point! With internal id it works fine... Thanks for the guidelines! :)

**Comment by amo13** [comment]
What is the format of the export that I get that way? I am trying to convert it to gpx but I don't know how. It seems the output format of the API call above is not GeoJSON at least...

**Comment by amo13** [comment]
I have tried a lot of different things, but I am still stuck with the exported file. It is apparently formatted in JSON, but it does not look like it could be easily converted to a real gpx file that could be imported elsewhere.
@tananaev , do you remember if you chose a certain format while doing the export feature or if you just put the data into a JSON format, but without specific parsable format?
Any hint is greatly appreciated!

**Comment by tananaev** [comment]
@amo13, you have to provide `Accept` header. Are you doing it?

**Comment by amo13** [comment]
Thank you for your answer,
I already have got the exported data as a 14mb file. Sublime text editor confirms that it is valid JSON. The beginning looks like this:
```
[
  {
    "id": 1,
    "attributes": {
      "battery": "24.0",
      "ip": "84.178.57.135",
      "distance": 0.0,
      "totalDistance": 0.0
    },
    "deviceId": 1,
    "type": null,
    "protocol": "osmand",
    "serverTime": "2017-10-20T15:50:16.000+0000",
    "deviceTime": "2017-10-19T00:24:17.000+0000",
    "fixTime": "2017-10-19T00:24:17.000+0000",
    "outdated": false,
    "valid": true,
    "latitude": 51.3124173,
    "longitude": 9.4667776,
    "altitude": 0.0,
    "speed": 0.0,
    "course": 0.0,
    "address": null,
    "accuracy": 0.0,
    "network": null
  },
  {
    "id": 4,
    "attributes": {
      "battery": "24.0",
      "ip": "84.178.57.135",
      "distance": 13.42,
      "totalDistance": 17.44
    },
    "deviceId": 1,
    "type": null,
```
My question is rather: is this some kind of standard format for geodata? I am looking for a way to convert it to GPX or something that can later be converted to GPX.

**Comment by tananaev** [comment]
That question doesn't make any sense because, as I mentioned earlier in this thread, GPX export is available through API.

**Comment by amo13** [comment]
Ok, so I followed the discussion above to find out the internal device ID and issued the same curl command than above to get the export and what I get is exactly what I showed in my last post. This is not GPX, but some sort of JSON, isn't it?
I tried to change the accept header from "Accept: application/json" to "Accept: application/gpx" but that only gives me a "HTTP 406 Not Acceptable - NotAcceptableException (...)" response.
Thank you for your help with this

**Comment by tananaev** [comment]
Media type for GPX is `application/gpx+xml`:

https://en.wikipedia.org/wiki/GPS_Exchange_Format

**Comment by amo13** [comment]
Thank you for pointing that out.
Unfortunately, it still get the 406 error with my curl call:
`curl -u a****y@mailbox.org -H "Accept: application/gpx+xml" "https://traccar.a****y.cc/api/reports/route?deviceId=1&from=2017-12-01T00:00:00Z&to=2017-12-31T00:00:00Z" > export.gpx`
What I don't understand is that it does work with the "application/json" accept header. What could I be doing wrong here?

**Comment by tananaev** [comment]
You have to use positions endpoint, not reports.

**Comment by amo13** [comment]
Awesome, now I was able to export successfully to GPX. Thank you very much for your hints!

Let me leave the curl command I used to successfully export to a export.gpx file in the current directory:
`curl -u user -H "Accept: application/gpx+xml" "https://traccar.yourdomain.com/api/positions?deviceId=1&from=2017-12-01T00:00:00Z&to=2017-12-31T00:00:00Z" > export.gpx`

Whoever is getting here while trying to find out how to do this might find this helpful. 

Don't forget to find out your (server-internal!) deviceId, which I did with this command:
`curl -u user -H "Accept: application/json" "https://traccar.yourdomain.com/api/devices"`

**Comment by tananaev** [comment]
Use `/api/positions/gpx` path instead.

**Comment by cnf** [comment]
> Use `/api/positions/gpx` path instead.

That did the trick! Thanks!
 Did I miss the documentation on this, or is this a recent change?

**Comment by tananaev** [comment]
Not very recent, but definitely more recent than this thread.

**Comment by cnf** [comment]
Aha, thanks! Maybe the docs should be refreshed to reflect this. The [swagger definitions](https://raw.githubusercontent.com/traccar/traccar/master/swagger.json) also don't know about this endpoint.

**Comment by tananaev** [comment]
Feel free to send a PR to update it.

**Comment by hax4dazy** [comment]
Is there any way to import GPX data? Moving over from dawarich to Traccar and wanted to import my data but I can't seem to find that option.

**Comment by tananaev** [comment]
Not possible currently.
